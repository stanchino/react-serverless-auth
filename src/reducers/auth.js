import { signUpRoutine, confirmRegistrationRoutine, signInRoutine, authRoutine, signOutRoutine, passwordResetRequestRoutine } from "../actions";

const flash = { flash: { error: null, notice: null } };

export const initialState = {
    loading: false,
    isLoggedIn: false,
    isRegistered: false,
    passwordResetRequested: false,
    profile: null,
    pathname: '/',
    user: null,
    ...flash
};

export default (state = initialState, action) => {
    switch (action.type) {
        case signUpRoutine.REQUEST:
        case confirmRegistrationRoutine.REQUEST:
        case signInRoutine.REQUEST:
        case signOutRoutine.REQUEST:
        case passwordResetRequestRoutine.REQUEST:
            return { ...state, loading: true, ...flash, ...action.payload };
        case authRoutine.REQUEST:
            return { ...state, loading: true };
        case signUpRoutine.SUCCESS:
        case signInRoutine.SUCCESS:
            return { ...state, isRegistered: true, ...action.payload };
        case passwordResetRequestRoutine.SUCCESS:
            return { ...state, passwordResetRequested: true, ...action.payload };
        case signUpRoutine.FAILURE:
            return { ...state, isRegistered: false };
        case authRoutine.SUCCESS:
            return { ...state, isLoggedIn: true, isRegistered: true, ...action.payload };
        case authRoutine.FAILURE:
            return { ...state, isLoggedIn: false, ...action.payload };
        case passwordResetRequestRoutine.FAILURE:
            return { ...state, passwordResetRequested: false, ...action.payload };
        case signUpRoutine.FULFILL:
        case confirmRegistrationRoutine.FULFILL:
        case signInRoutine.FULFILL:
        case authRoutine.FULFILL:
        case passwordResetRequestRoutine.FULFILL:
            return { ...state, loading: false };
        case signOutRoutine.FULFILL:
            return initialState;
        default:
            return state;
    }
};