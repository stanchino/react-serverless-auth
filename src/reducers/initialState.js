export const flash = { flash: { error: null, notice: null } };

export const initialState = {
  loading: false,
  isLoggedIn: false,
  isRegistered: false,
  passwordResetRequested: false,
  profile: null,
  pathname: '/',
  user: null,
  ...flash
};