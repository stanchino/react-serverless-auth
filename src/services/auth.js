import { currentUser } from "./utils"

export default () => (
    new Promise((resolve, reject) => {
        if (null === currentUser) {
            resolve(null);
        }
        currentUser.getSession(error => {
            if (error) {
                reject(error);
            } else {
                resolve(currentUser);
            }
        });
    })
);