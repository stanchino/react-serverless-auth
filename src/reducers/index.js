import { reducer as formReducer  } from "redux-form";
import auth from "./auth";

export default {
    auth: auth,
    form: formReducer
};