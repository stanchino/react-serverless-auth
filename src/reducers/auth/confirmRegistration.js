import { confirmRegistrationRoutine } from "../../actions";

import { initialState, flash } from "../initialState";

export default (state = initialState, action) => {
  switch (action.type) {
    case confirmRegistrationRoutine.REQUEST:
      return { ...state, loading: true, ...flash, ...action.payload };
    case confirmRegistrationRoutine.FULFILL:
      return { ...state, loading: false };
    default:
      return state;
  }
};