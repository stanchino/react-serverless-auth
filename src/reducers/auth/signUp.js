import { signUpRoutine } from "../../actions";

import { initialState, flash } from "../initialState";

export default (state = initialState, action) => {
  switch (action.type) {
    case signUpRoutine.REQUEST:
      return { ...state, loading: true, ...flash, ...action.payload };
    case signUpRoutine.SUCCESS:
      return { ...state, isRegistered: true, ...action.payload };
    case signUpRoutine.FAILURE:
      return { ...state, isRegistered: false };
    case signUpRoutine.FULFILL:
      return { ...state, loading: false };
    default:
      return state;
  }
};