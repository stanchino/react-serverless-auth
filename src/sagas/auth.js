import {call, put, select } from "redux-saga/effects";
import { authRoutine } from "../actions";
import { authRequest, userAttributes } from "../services";

export const getUser = state => (state.auth.user);

export function* handleAuthSaga() {
    let profile;
    let user = yield select(getUser);
    try {
        yield put(authRoutine.request());
        if (null === user)  {
            user = yield call(authRequest);
        }
        if (user) {
            profile = yield call(userAttributes, user);
            yield put(authRoutine.success({ user: user, profile: profile }));
        }
    } catch (error) {
        yield put(authRoutine.failure({ flash: { error: error.message } }));
    } finally {
        yield put(authRoutine.fulfill())
    }
}