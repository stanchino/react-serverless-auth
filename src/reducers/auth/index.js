import reduceReducers from "reduce-reducers";

import auth from "./auth";
import confirmRegistration from "./confirmRegistration";
import passwordReset from "./passwordReset";
import signIn from "./signIn";
import signOut from "./signOut";
import signUp from "./signUp";

export default reduceReducers(auth, confirmRegistration, passwordReset, signIn, signOut, signUp);