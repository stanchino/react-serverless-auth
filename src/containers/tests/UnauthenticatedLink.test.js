import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import createStore from "redux-mock-store";
import { routerMiddleware, ConnectedRouter } from "react-router-redux";
import createHistory from "history/createHashHistory";
import { Switch, Route } from "react-router";
import { mount } from "enzyme";

import { initialState } from "../../reducers/auth";

import { UnauthenticatedLink } from "..";

const history = createHistory();
const store = state => createStore([routerMiddleware(history)])({ auth: state });
const app = state => (
    <Provider store={store(state)}>
        <ConnectedRouter history={history}>
            <div>
                <nav>
                    <UnauthenticatedLink to={"/auth"}>Link</UnauthenticatedLink>
                </nav>
                <Switch>
                    <Route path={"/"} exact><div>Home</div></Route>
                    <Route path={"/auth"}><div>Auth</div></Route>
                </Switch>
            </div>
        </ConnectedRouter>
    </Provider>
);

describe("UnauthenticatedLink Component", () => {
    it("matches the snapshot", () => {
        expect(renderer.create(app(initialState)).toJSON()).toMatchSnapshot();
    });

    it("while page is loading", () => {
        expect(mount(app({ loading: true })).find('NavLink')).toHaveLength(0);
    });

    describe("when the state is loaded", () => {
        const state = isLoggedIn => ({ loading: false, isLoggedIn: isLoggedIn });
        const expectedLink = isLoggedIn => (expect(mount(app(state(isLoggedIn))).find('NavLink').length));

        it("when user is signed in is loaded", () => expectedLink(true).toEqual(0));
        it("when user is not signed in is loaded", () => expectedLink(false).toEqual(1));
    });
});