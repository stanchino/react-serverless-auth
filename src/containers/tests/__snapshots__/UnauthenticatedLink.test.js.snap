// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnauthenticatedLink Component matches the snapshot 1`] = `
<div>
  <nav>
    <a
      aria-current={false}
      className={undefined}
      href="#/auth"
      onClick={[Function]}
      style={undefined}
    >
      Link
    </a>
  </nav>
  <div>
    Home
  </div>
</div>
`;
